### PR TITLE
[patch] add pprof endpoint

### PR DESCRIPTION
**Pull request by kevindiu**

### CODE REVIEW

**Comment by WindzCUHK** [comment]
need a flag to enable profiling separate?

**Comment by kevindiu** [comment]
> LGTM
> 
> need a flag to enable profiling separate?

thinking we shouldn't expose profiling interface in production environment so i just put it in debug server. what do you think?

**Comment by kevindiu** [comment]
@WindzCUHK please review latest commits and comment :)

File: config/config.go [hunk 1]
```diff
@@ -164,6 +161,21 @@ type Authorization struct {
...
+// DebugServer represents the server for debug use.
+type DebugServer struct {
+	// Enable represents if user want to enable debug server functionality.
+	Enable bool `yaml:"enable"`
+
+	// Port represents debug server port.
+	Port int `yaml:"port"`
+
+	// EnableDump represents if user want to enable memory dump functionality.
+	EnableDump bool `yaml:"enable_dump"`
```
Comment (WindzCUHK) [inline config/config.go]: `EnableDump` => `EnablePolicyDump`?

File: router/debug_router.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ func NewDebugRouter(cfg config.Server, a service.Authorizationd) *http.ServeMux
 		dur = time.Second * 3
 	}
 
-	for _, route := range NewDebugRoutes(cfg, a) {
+	for _, route := range NewDebugRoutes(cfg.DebugServer, a) {
```
Comment (WindzCUHK) [inline router/debug_router.go]: cleaner to split the routes to different function?
```go
if cfg.DebugServer.EnablePolicyDump {
    for _, route := range NewPolicyDumpRoutes(a) {}
}
if cfg.DebugServer.EnableProfiling {
    for _, route := range NewPprofRoutes() {}
}
```

File: config/config.go (hunk 1 above)
Comment (kevindiu) [reply config/config.go]: For future if we add more cache then another flag need to be added, so I think enable dump should be better.

File: router/debug_router.go (hunk 2 above)
Comment (kevindiu) [reply router/debug_router.go]: I think it is ok to put in one function, maybe only separate the under-layer logic?

Comment (kevindiu) [reply router/debug_router.go]: The logic is very simple, and I will keep the current implementation.

**Comment by kevindiu** [comment]
LGTM for the add comments commit.
For merge commit I think we do not need it?

**Comment by WindzCUHK** [comment]
> LGTM for the add comments commit.
> For merge commit I think we do not need it?

git problems... Please ignore that commit.

add test for coverage: https://github.com/yahoojapan/authorization-proxy/pull/29/commits/2dd061e96ec1a8bb5673062ce049a3583595d958
plx review

**Comment by kevindiu** [comment]
Maybe you use git merge instead of rebase?

**Comment by kevindiu** [comment]
I cannot trace if you changed any logic or not....

**Comment by kevindiu** [comment]
For latest test is LGTM other than the merge commit.

**Comment by WindzCUHK** [comment]
> I cannot trace if you changed any logic or not....

git UI has bug...
select commit by URL directly

https://github.com/yahoojapan/authorization-proxy/pull/29/files/227f75c484a72d5b106f609f6e781c47a8c1ee5f..2dd061e96ec1a8bb5673062ce049a3583595d958

**Comment by kevindiu** [comment]
I think it is not UI bug... If I only select the merge commit then all changes in this PR will included into 1 commit... I will try to force push it.
https://github.com/yahoojapan/authorization-proxy/pull/29/commits/3ba9c5829cb3327c7e83a3f9e61ac408e563ad08

**Comment by WindzCUHK** [comment]
LGTM

**Pull request merged**